src/commands/restartServer.ts

```diff
@@ -0,0 +1,51 @@
+import {
+  ShutdownRequest,
+  ExitNotification
+} from "vscode-languageserver-protocol";
+import { LanguageClient } from "../interfaces/LanguageClient";
+import { exec, ChildProcess } from "child_process";
+import { ShowMessage, normalize } from "../interfaces/ShowMessage";
+
+const timeoutMs = 4000;
+
+/**
+ * Try to gracefully shutdown the server using LSP `shutdown` and `exit` endpoints.
+ * If the server is unresponsive, fall back to killing its process.
+ *
+ * @param client
+ * @param workspace
+ */
+export function restartServer(
+  client: LanguageClient,
+  showMessage: ShowMessage
```

Comment:
here's a couple of examples of what I mentioned in the PR description.

There's no common `LanguageClient` interface between VS Code and coc.nvim, so we create one ourself and the clients can just pass in their respective implementations (as long as the types align we're fine, since TypeScript has a structural type system).

`ShowMessage` is a bit more complicated, since VS Code and coc.nvim have a different API for showing a message to the user. We abstract their differences away in `ShowMessage` and use a `normalize` function to unify the two APIs.